### PR TITLE
Integrate upstream incremental index update optimizations and other improvements

### LINKED ISSUE #5809: Unpacking index is extremely slow
### Apache NetBeans version

Apache NetBeans 17

### What happened

Related bug: https://bz.apache.org/netbeans/show_bug.cgi?id=239915

I have the default value for updating maven index frequency of one week.
Every time it gets updated, the step "unpacking index" takes a lot of time - about 20 minutes. Also the fans are running at full speed.

### How to reproduce

Wait for the next weekly maven index update.

### Did this work correctly in an earlier version?

No / Don't know

### Operating System

Ubuntu 22.10

### JDK

OpenJDK 20

### Apache NetBeans packaging

Apache NetBeans binary zip

### Anything else

_No response_

### Are you willing to submit a pull request?

No

### PR DESCRIPTION
**Pull request by mbien**
maven-indexer 7.0.3 contains two optimizations:
 - https://github.com/apache/maven-indexer/pull/318
 - https://github.com/apache/maven-indexer/pull/317
 
This PR has also some code cleanup and deprecation fixes, improvements to logging and makes CRC computation in ClassDependencyIndexCreator thread safe.

```
INFO [org.netbeans.modules.maven.indexer.NexusRepositoryIndexerImpl]: Indexing [scan] of local took 50.88s.
INFO [org.netbeans.modules.maven.indexer.NexusRepositoryIndexerImpl]: Indexing [incremental download, update] of central took 69.22s.
```

closes #5809

### CODE REVIEW

**Review by matthiasblaesing — COMMENTED** [review]
Looks good apart from the API breaks, please fix and then I think this is good to go.

File: java/maven.indexer/src/org/netbeans/modules/maven/indexer/api/NBArtifactInfo.java [hunk 1]
```diff
@@ -40,7 +40,7 @@ public NBArtifactInfo(String name) {
         this.name = name;
     }
 
-    public boolean removeVersionInfo(Object o) {
+    public boolean removeVersionInfo(NBVersionInfo o) {
```
[the remark this answers is not in the document]
Comment (mbien) [reply java/maven.indexer/src/org/netbeans/modules/maven/indexer/api/NBArtifactInfo.java:43]: i think the whole module is only exposed to friend dependencies. So dependent modules outside of NB would have to rebuild with every NB release anyway, right? Does the major version increase do something or is this only a convention?

[the remark this answers is not in the document]
Comment (matthiasblaesing) [reply java/maven.indexer/src/org/netbeans/modules/maven/indexer/api/NBArtifactInfo.java:43]: From my POV: If a module declares an API, we need to be correct with it. NetBeans has a level for what you describe: It is an implementation dependency. These are hard-tied to the corresponding NetBeans build.

So I see outs here: Bump major or just leave it as is. It is a minor problem, not worth breaking API for.

[the remark this answers is not in the document]
Comment (mbien) [reply java/maven.indexer/src/org/netbeans/modules/maven/indexer/api/NBArtifactInfo.java:43]: ok going to revert the change

**Pull request merged**